diff --git a/yate/configure.ac b/yate/configure.ac
index 3ec58ad..b377651 100644
--- a/yate/configure.ac
+++ b/yate/configure.ac
@@ -1798,7 +1798,8 @@ AC_CONFIG_FILES([packing/rpm/yate.spec
                  share/sounds/Makefile
                  share/help/Makefile
                  share/data/Makefile
-                 conf.d/Makefile])
+                 conf.d/Makefile
+                 tools/Makefile])
 AC_CONFIG_FILES([yate-config],[chmod +x yate-config])
 AC_CONFIG_FILES([run],[chmod +x run])
 CONFIGURE_FILES=`echo "$ac_config_files config.status config.log" | sed 's,packing/[[^ ]]* *,,g'`
