diff --git a/yatebts/sar/Makefile.in b/yatebts/sar/Makefile.in
index 0f4beb2..81c3e3c 100644
--- a/yatebts/sar/Makefile.in
+++ b/yatebts/sar/Makefile.in
@@ -4,7 +4,7 @@
 # override DESTDIR at install time to prefix the install directory
 DESTDIR :=
 
-SCRIPTS := searchandrescue.js.old sar_lib.js searchandrescue.js
+SCRIPTS := sar_lib.js searchandrescue.js
 
 prefix = @prefix@
 exec_prefix = @exec_prefix@
