diff --git a/yatebts/sar/sar_lib.js b/yatebts/sar/sar_lib.js
index 1b72d02..4c1259d 100644
--- a/yatebts/sar/sar_lib.js
+++ b/yatebts/sar/sar_lib.js
@@ -64,10 +64,17 @@ function imsiPermitted(imsi) {
       || (!testing && !forbiddenImsis[imsi]);
 }
 
-function writeToOCP(obj) {
-   var objStr = JSON.stringify(obj);
-   Engine.debug(Engine.DebugInfo, "Writing to OCP: " + objStr);
-   sar.writeToOCP(objStr);
+function writeToOCP(type, data) {
+   var now = Date.now() / 1000;
+   var ocpMsg = {
+      'time': now,
+      'type': type,
+      'data': data
+   };
+
+   var ocpMsgStr = JSON.stringify(ocpMsg);
+   Engine.debug(Engine.DebugInfo, "Writing to OCP: " + ocpMsgStr);
+   sar.writeToOCP(ocpMsgStr);
 }
 
 /* ############### SAR Initialization ############### */
@@ -587,12 +594,7 @@ function onPhyinfo(msg) {
       'DnRSSIdBm': msg.DnRSSIdBm,
       'time': msg.time
    };
-   writeToOCP({
-      'type': 'phy',
-      'IMSI': subscriber.imsi,
-      'TMSI': subscriber.tmsi,
-      'data': phyinfo
-   });
+   writeToOCP('phy', phyinfo);
 
    // update the subscriber's phyinfo
    subscriber.lastPhyinfo = phyinfo;
