diff --git a/yatebts/sar/subscribers.js b/yatebts/sar/subscribers.js
index 4755fde..721f114 100644
--- a/yatebts/sar/subscribers.js
+++ b/yatebts/sar/subscribers.js
@@ -118,7 +118,7 @@ function imsiPermitted(imsi) {
 
 function getSubscriberByMSISDN(msisdn) {
    for (var i = 0; i < activeSubscribers.length; ++i) {
-      if (activeSubscribers[i].msisdn === msisdn)
+      if (activeSubscribers[i].msisdn == msisdn)
          return activeSubscribers[i];
    }
    return null;
