diff --git a/yatebts/sar/sar.conf.sample b/yatebts/sar/sar.conf.sample
index 2e39c00..452995e 100644
--- a/yatebts/sar/sar.conf.sample
+++ b/yatebts/sar/sar.conf.sample
@@ -43,13 +43,13 @@ helloText=phone detected
 ; For best performance, choose intervals with a large (>1000) greatest common divisor 
 ; as a master periodic action will run at that interval. 
 ; 
-; NOTE: untested with values less than 1000
+; NOTE: untested with values less than 500
 
 ; How often a heartbeat status message is sent to the OCP
 ;heartbeat=10000
 
 ; How often cell phone signal strength is measured.
-;phyinfoPolling=3000
+;phyinfoPolling=1000
 
 ; How often cell subscriptions are checked for expiration.
 ;subscriberExpire=5000
diff --git a/yatebts/sar/sar_config.js b/yatebts/sar/sar_config.js
index 256b025..ae39a99 100644
--- a/yatebts/sar/sar_config.js
+++ b/yatebts/sar/sar_config.js
@@ -29,12 +29,12 @@ var config = {
 
 // intervals (in ms) between different runs
 var intervals = {
-   // 'master': , // calculated when loadConfiguration is run to be the gcd of all other timers
+   // 'master': null, // calculated when loadConfiguration is run to be the gcd of all other timers
    'heartbeat': 10000,
-   'phyinfoPolling': 3000,
-   'subscriberExpire': 5000,
+   'phyinfoPolling': 1000,
+   'subscriberExpire': 60000,
    'onInterval': 10000,
-   'sendSMS': 1000
+   'sendSMS': 100
 };
 
 var helloText = "phone detected";  // sent to a phone when they join the network
diff --git a/yatebts/sar/utils.js b/yatebts/sar/utils.js
index 0822757..6017825 100644
--- a/yatebts/sar/utils.js
+++ b/yatebts/sar/utils.js
@@ -115,6 +115,7 @@ function gcdNums(nums) {
    return gcd;
 }
 
+// calculate the GCD of 2 numbers with Euclid's Algorithm
 function gcdTwoNums(a, b) {
    if (a < b) {
       var tmp = a;
